validate.go

```diff
@@ -48,38 +48,36 @@ func validateRegistrations(regs []*codegen.Registration) error {
 				if _, ok := intfs[v.Type]; !ok {
 					// T is not a registered component interface.
 					err := fmt.Errorf(
-						"component implementation struct %v has field %v, but component %v was not registered; maybe you forgot to run 'weaver generate'",
+						"component implementation struct %v has component reference field %v, but component %v was not registered; maybe you forgot to run 'weaver generate'",
 						reg.Impl, f.Type, v.Type,
 					)
 					errs = append(errs, err)
 				}
 
 			case f.Type == reflection.Type[Listener]():
 				// f is a weaver.Listener.
-				if tag, ok := f.Tag.Lookup("weaver"); ok && !isValidListenerName(tag) {
-					err := fmt.Errorf("component implementation struct %v has invalid listener tag %q", reg.Impl, tag)
+				name := f.Name
+				if tag, ok := f.Tag.Lookup("weaver"); ok {
+					if !token.IsIdentifier(tag) {
+						err := fmt.Errorf("component implementation struct %v has invalid listener tag %q", reg.Impl, tag)
+						errs = append(errs, err)
+						continue
+					}
+					name = tag
+				}
+				found := false
+				for _, lis := range reg.Listeners {
+					if lis == name {
+						found = true
+						break
+					}
+				}
+				if !found {
+					err := fmt.Errorf("component implementation struct %v has a listener field %v, but listener %v hasn't been registered; maybe you forgot to run 'weaver generate'", reg.Impl, name, name)
 					errs = append(errs, err)
 				}
 			}
 		}
 	}
 	return errors.Join(errs...)
 }
-
-// isValidListenerName returns whether the provided name is a valid
```

Comment:
I replaced `token.IsIdentifier` with `isValidListenerName` to remove the dependency on the `token` package, as recommended by Sanjay here: https://github.com/ServiceWeaver/weaver/pull/473#discussion_r1268593783. Happy to revisit the decision and switch back to using `token`, but just wanted to give the full context.